immer/detail/hamts/bits.hpp

```diff
@@ -18,43 +18,91 @@ namespace immer {
 namespace detail {
 namespace hamts {
 
-using bits_t   = std::uint32_t;
-using bitmap_t = std::uint32_t;
-using count_t  = std::uint32_t;
-using shift_t  = std::uint32_t;
 using size_t   = std::size_t;
 using hash_t   = std::size_t;
 
-template <bits_t B, typename T=count_t>
+template <size_t B>
+struct types
+{
+    static_assert(B < 6, "B > 6 is not supported.");
+
+    using bits_t   = std::uint32_t;
+    using bitmap_t = std::uint32_t;
+    using count_t  = std::uint32_t;
+    using shift_t  = std::uint32_t;
+};
+
+template <>
+struct types<6>
+{
+    using bits_t   = std::uint64_t;
+    using bitmap_t = std::uint64_t;
+    using count_t  = std::uint64_t;
+    using shift_t  = std::uint64_t;
+};
+
+template <size_t B>
+using bits_t   = types<B>::bits_t;
+
+template <size_t B>
+using bitmap_t = types<B>::bitmap_t;
+
+template <size_t B>
+using count_t  = types<B>::count_t;
+
+template <size_t B>
+using shift_t  = types<B>::shift_t;
+
+template <size_t B, typename T=count_t<B>>
 constexpr T branches = T{1} << B;
 
-template <bits_t B, typename T=size_t>
+template <size_t B, typename T=size_t>
 constexpr T mask = branches<B, T> - 1;
 
-template <bits_t B, typename T=count_t>
+template <size_t B, typename T=count_t<B>>
 constexpr T max_depth = (sizeof(hash_t) * 8 + B - 1) / B;
 
-template <bits_t B, typename T=count_t>
-constexpr T max_shift = max_depth<B, count_t> * B;
+template <size_t B, typename T=count_t<B>>
+constexpr T max_shift = max_depth<B, count_t<B>> * B;
 
 #define IMMER_HAS_BUILTIN_POPCOUNT 1
 
-inline count_t popcount(bitmap_t x)
+template <size_t B>
+inline auto popcount_fallback(bitmap_t<B> x)
 {
-#if IMMER_HAS_BUILTIN_POPCOUNT
-#  if defined(_MSC_VER)
-    return  __popcnt(x);
-#  else
-    return __builtin_popcount(x);
-#  endif
-#else
     // More alternatives:
     // https://en.wikipedia.org/wiki/Hamming_weight
     // http://wm.ite.pl/articles/sse-popcount.html
     // http://graphics.stanford.edu/~seander/bithacks.html#CountBitsSetParallel
     x = x - ((x >> 1) & 0x55555555);
     x = (x & 0x33333333) + ((x >> 2) & 0x33333333);
     return ((x + (x >> 4) & 0xF0F0F0F) * 0x1010101) >> 24;
+}
+
+template <size_t B>
+inline count_t<B> popcount(bitmap_t<B> x)
+{
+#if IMMER_HAS_BUILTIN_POPCOUNT
+#  if defined(_MSC_VER)
+    return  __popcnt(x);
+#  else
+    return __builtin_popcount(x);
+#  endif
+#else
+    return popcount_fallback(x);
+#endif
+}
+
+inline count_t<6> popcount(bitmap_t<6> x)
```

Comment:
Instead of overloading popcount for different `bitmap_t<X>`, I would suggest defining the overloads in terms of the underlying representation types. This is, to provide functions:
```
std::uint32_t popcount(std::uint32_t x);
std::uint64_t popcount(std::uint64_t x);
...
```

Reply:
Otherwise you are unnecesarily coupling a general function (_popcount_) to implementation details of hamts (the notion of _bitmap_). Does this make sense?

Reply:
Good suggestion, thanks!